src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs

```diff
@@ -318,5 +327,24 @@ private async Task<IEnumerable<Project>> GetProjectsInSolutionAsync(DTE dte)
             IEnumerable<Project> supportedProjects = allProjects.Where(EnvDTEProjectUtility.IsSupported);
             return supportedProjects;
         }
+
+        public bool TryCreateUserWideContext(out IVsPathContext vsPathContext)
+        {
+            // invoke async operation from within synchronous method
+            vsPathContext = NuGetUIThreadHelper.JoinableTaskFactory.Run(() => TryCreateUserWideContextAsync());
```

Comment:
Would it be worthwhile throwing if we're on the UI thread, to make sure callers call us on a background thread?

Although if a C++ component is calling us through COM, I don't know if that's possible.

Reply:
You can always explicitly switch to background thread:

await TaskScheduler.Default;

---
In reply to: [446392155](https://github.com/NuGet/NuGet.Client/pull/3478#discussion_r446392155) [](ancestors = 446392155)

Reply:
I'll add a comment that this API can be safely invoked on a background thread. 
The threading logic is similar as the other IVSPathContextProvider methods.

Reply:
I'd like to leave this as is. 

> Would it be worthwhile throwing if we're on the UI thread, to make sure callers call us on a background thread?

This seems like an anti-pattern to me :) 

I documented the threading affinity.